kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java

```diff
@@ -0,0 +1,184 @@
+  }
+
+
+  @Test
+  public void should_be_able_to_return_service_account_name_when_impersonating_current_user() {
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(SERVICE_ACCOUNT);
+    requestConfig.setImpersonateGroups("system:authenticated", "system:authenticated:oauth");
+
+    User user = client.currentUser();
+    assertThat(user.getMetadata().getName()).isEqualTo(SERVICE_ACCOUNT);
+  }
+
+
+  @Test
+  public void should_be_able_to_create_a_project_impersonating_service_account() {
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(SERVICE_ACCOUNT);
+    requestConfig.setImpersonateGroups("system:authenticated", "system:authenticated:oauth");
+    // Create a project
+    ProjectRequest projectRequest = client.projectrequests().createNew()
+      .withNewMetadata()
+      .withName(NEW_PROJECT)
+      .endMetadata()
+      .done();
+
+    // Grab the requester annotation
+    String requester = projectRequest.getMetadata().getAnnotations().get("openshift.io/requester");
+    assertThat(requester).isEqualTo(SERVICE_ACCOUNT);
+  }
+
+
+  @After
+  public void cleanup() {
+    // Reset original authentication
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(null);
+    requestConfig.setImpersonateGroups(null);
+
+    // Delete Cluster Role
+    client.rbac().kubernetesClusterRoles().inNamespace(currentNamespace).delete(impersonatorRole);
+    await().atMost(1, TimeUnit.MINUTES).until(
+      new Callable<Boolean>() {
+        @Override
+        public Boolean call() {
+          return client.rbac().kubernetesClusterRoles().inNamespace(currentNamespace).withName("impersonator").get() == null;
+        }
+      }
+    );
+    // Delete Cluster Role binding
+    client.rbac().kubernetesClusterRoleBindings().inNamespace(currentNamespace).delete(impersonatorRoleBinding);
+    await().atMost(1, TimeUnit.MINUTES).until(
+      new Callable<Boolean>() {
+        @Override
+        public Boolean call() {
+          return client.rbac().kubernetesClusterRoleBindings().inNamespace(currentNamespace).withName("impersonator-role").get() == null;
+        }
+      }
+    );
+    // Delete project
+    client.projects().withName(NEW_PROJECT).delete();
+    await().atMost(1, TimeUnit.MINUTES).until(
+      new Callable<Boolean>() {
+        @Override
+        public Boolean call() {
+          return client.projects().withName(NEW_PROJECT).get() == null;
+        }
+      }
+    );
+    // Delete ServiceAccounts
+    client.serviceAccounts().inNamespace(currentNamespace).delete();
+    await().atMost(1, TimeUnit.MINUTES).until(
+      new Callable<Boolean>() {
+        @Override
+        public Boolean call() {
+          return client.serviceAccounts().list().getItems().isEmpty();
```

Comment:
How about making those anonymous classes methods named after the comments? 

`await().atMost(1, TimeUnit.MINUTES).until(serviceAccountIsDeleted())`?

Also - do we have to wait for all the steps? What is we e.g. make project name "randomized"?

Reply:
Done.  Let' s keep it waiting for all the steps for now, as it's safer this way